File: 5-Problem-A-Day/245. Shortest Word Distance III.java
class Solution {
    public int shortestWordDistance(String[] words, String word1, String word2) {
        int i1 = -1, i2 = -1, min = words.length;
        boolean update1 = true;
        for (int i = 0; i < words.length; i++) {
            if (words[i].equals(word1) && (update1 || !words[i].equals(word2))) {
                i1 = i;
                update1 = false;
            } else if (words[i].equals(word2) && (!update1 || !words[i].equals(word1))) {
                i2 = i;
                update1 = true;
            }
            if (i1 != -1 && i2 != -1 && i1 != i2)
                min = Math.min(min, Math.abs(i1 - i2));
        }
        return min;
    }
}

File: 2021/245. Shortest Word Distance III.cpp
class Solution {
public:    
    int shortestWordDistance(vector<string>& words, string word1, string word2) {
        int index1 = -1, index2 = -1, res = words.size();
        for (int i = 0; i < words.size(); i++) {
            if (word1 != word2) {
                if (words[i] == word1) {
                    index1 = i;
                } 
                if (words[i] == word2) {
                    index2 = i;
                }
            } else {
                if (words[i] == word1) {
                    if (index1 > index2) {
                        index2 = i;
                    } else {
                        index1 = i;
                    }
                }
            }            
            if (index1 != -1 && index2 != -1) {
                res = min(res, abs(index1 - index2));
            }
        }
        return res;
    }
};

class Solution {
public:    
    int shortestWordDistance(vector<string>& words, string word1, string word2) {
        int index1 = -1, index2 = -1, res = words.size();
        for (int i = 0; i < words.size(); i++) {
            if (words[i] == word1 && (word1 != word2 || index1 <= index2)) {
                index1 = i;
            } else if (words[i] == word2) {
                index2 = i;
            }                                 
            if (index1 != -1 && index2 != -1) {
                res = min(res, abs(index1 - index2));
            }
        }
        return res;
    }
};